core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -157,22 +189,15 @@ class AvroUpsertMonitorActor(
     case StartMonitorPass =>
       // start the process by pulling a message and sending it back to self
       pubSubDao.pullMessages(pubSubSubscriptionName, 1).map(_.headOption) pipeTo self
+      arrowPubSubDAO.pullMessages(arrowPubSubSubscriptionName, 1).map(_.headOption) pipeTo self
```

Comment:
At every pass, we pull messages from both topics.

Reply:
How does this work? Does `pipeTo` place a message in the actor mailbox with the result of the computation or something? I'm trying to understand what happens if there's a message on both queues, since this is a mix of "two things sequentially" and "things that involve futures / actor messages", and I don't know the answer. Maybe a test to confirm that we process both messages if there's one on each pubsub?

Reply:
I tested by doing the following:

- Changed the config values for `pollInterval` and `pollJitter` so the monitor pass happens less frequently
- Added logging so that the actor logs when each monitor pass happens and at what time
- Stood up front and back rawls locally
- Called the old CF path by hitting the dev orch endpoint and the called the new import service path by making curl calls. Hit both of these endpoints many times very quickly (several times a second)
- Watched logs - looking at each block of logging between monitor passes

What I saw was that some of the monitor passes do receive messages from both the topics

`- MONITOR PASS: 2020-02-14T17:51:56
rawls [DEBUG] [17:51:56.744] [scala-execution-context-global-630] o.b.d.r.google.HttpGooglePubSubDAO - {"method":"POST","payload":{"maxMessages":1,"returnImmediately":true},"statusCode":200,"time_ms":510,"url":"https://pubsub.googleapis.com/v1/projects/terra-arrow-dev/subscriptions/avro-import-topic-sub:pull"}
rawls [INFO] [17:51:56.762] [rawls-akka.actor.default-dispatcher-15] o.b.d.r.m.AvroUpsertMonitorActor - received async upsert message: PubSubMessage(ISE-MD5FU0RQBhYsXUZIUTcZCGhRDk9eIz81IChFEAEIFAV8fXBBXnVeXHJoUQ0Zcn1gcmpeF1UBEFUrW1EbB2hObXUlbgQRenRkdWNeFgIDQ1R2V3OYnIHvq_ePYBclSvHt3qhvM-qe2r5MZiU9XxJLLD5-PTNFQV5AEkw2BkRJUytDCypYEU4E,{
rawls   "kind": "storage#object",
rawls   "id": "avro-translated-json-dev/62738f6f-2967-4257-90df-0f4213549946/metadata.json/1581702702699277",
rawls   "selfLink": "https://www.googleapis.com/storage/v1/b/avro-translated-json-dev/o/62738f6f-2967-4257-90df-0f4213549946%2Fmetadata.json",
rawls   "name": "62738f6f-2967-4257-90df-0f4213549946/metadata.json",
rawls   "bucket": "avro-translated-json-dev",
rawls   "generation": "1581702702699277",
rawls   "metageneration": "1",
rawls   "contentType": "application/json",
rawls   "timeCreated": "2020-02-14T17:51:42.699Z",
rawls   "updated": "2020-02-14T17:51:42.699Z",
rawls   "storageClass": "STANDARD",
rawls   "timeStorageClassUpdated": "2020-02-14T17:51:42.699Z",
rawls   "size": "225",
rawls   "md5Hash": "1urPdSavDj60DDVz86f3bQ==",
rawls   "mediaLink": "https://www.googleapis.com/download/storage/v1/b/avro-translated-json-dev/o/62738f6f-2967-4257-90df-0f4213549946%2Fmetadata.json?generation=1581702702699277&alt=media",
rawls   "crc32c": "vWdymg==",
rawls   "etag": "CI2m8N3N0ecCEAE="
rawls }
rawls ,Map(eventType -> OBJECT_FINALIZE, bucketId -> avro-translated-json-dev, eventTime -> 2020-02-14T17:51:42.699114Z, objectId -> 62738f6f-2967-4257-90df-0f4213549946/metadata.json, payloadFormat -> JSON_API_V1, notificationConfig -> projects/_/buckets/avro-translated-json-dev/notificationConfigs/3, objectGeneration -> 1581702702699277))
rawls [INFO] [17:51:56.774] [rawls-akka.actor.default-dispatcher-15] o.b.d.r.m.AvroUpsertMonitorActor - found ignorable file metadata.json message for job 62738f6f-2967-4257-90df-0f4213549946 with ackId ISE-MD5FU0RQBhYsXUZIUTcZCGhRDk9eIz81IChFEAEIFAV8fXBBXnVeXHJoUQ0Zcn1gcmpeF1UBEFUrW1EbB2hObXUlbgQRenRkdWNeFgIDQ1R2V3OYnIHvq_ePYBclSvHt3qhvM-qe2r5MZiU9XxJLLD5-PTNFQV5AEkw2BkRJUytDCypYEU4E in message: PubSubMessage(ISE-MD5FU0RQBhYsXUZIUTcZCGhRDk9eIz81IChFEAEIFAV8fXBBXnVeXHJoUQ0Zcn1gcmpeF1UBEFUrW1EbB2hObXUlbgQRenRkdWNeFgIDQ1R2V3OYnIHvq_ePYBclSvHt3qhvM-qe2r5MZiU9XxJLLD5-PTNFQV5AEkw2BkRJUytDCypYEU4E,{
rawls   "kind": "storage#object",
rawls   "id": "avro-translated-json-dev/62738f6f-2967-4257-90df-0f4213549946/metadata.json/1581702702699277",
rawls   "selfLink": "https://www.googleapis.com/storage/v1/b/avro-translated-json-dev/o/62738f6f-2967-4257-90df-0f4213549946%2Fmetadata.json",
rawls   "name": "62738f6f-2967-4257-90df-0f4213549946/metadata.json",
rawls   "bucket": "avro-translated-json-dev",
rawls   "generation": "1581702702699277",
rawls   "metageneration": "1",
rawls   "contentType": "application/json",
rawls   "timeCreated": "2020-02-14T17:51:42.699Z",
rawls   "updated": "2020-02-14T17:51:42.699Z",
rawls   "storageClass": "STANDARD",
rawls   "timeStorageClassUpdated": "2020-02-14T17:51:42.699Z",
rawls   "size": "225",
rawls   "md5Hash": "1urPdSavDj60DDVz86f3bQ==",
rawls   "mediaLink": "https://www.googleapis.com/download/storage/v1/b/avro-translated-json-dev/o/62738f6f-2967-4257-90df-0f4213549946%2Fmetadata.json?generation=1581702702699277&alt=media",
rawls   "crc32c": "vWdymg==",
rawls   "etag": "CI2m8N3N0ecCEAE="
rawls }
rawls ,Map(eventType -> OBJECT_FINALIZE, bucketId -> avro-translated-json-dev, eventTime -> 2020-02-14T17:51:42.699114Z, objectId -> 62738f6f-2967-4257-90df-0f4213549946/metadata.json, payloadFormat -> JSON_API_V1, notificationConfig -> projects/_/buckets/avro-translated-json-dev/notificationConfigs/3, objectGeneration -> 1581702702699277))
rawls [INFO] [17:51:56.774] [rawls-akka.actor.default-dispatcher-15] o.b.d.r.m.AvroUpsertMonitorActor - acking message with ackId ISE-MD5FU0RQBhYsXUZIUTcZCGhRDk9eIz81IChFEAEIFAV8fXBBXnVeXHJoUQ0Zcn1gcmpeF1UBEFUrW1EbB2hObXUlbgQRenRkdWNeFgIDQ1R2V3OYnIHvq_ePYBclSvHt3qhvM-qe2r5MZiU9XxJLLD5-PTNFQV5AEkw2BkRJUytDCypYEU4E
rawls [DEBUG] [17:51:56.875] [scala-execution-context-global-632] o.b.d.r.google.HttpGooglePubSubDAO - {"method":"POST","payload":{"maxMessages":1,"returnImmediately":true},"statusCode":200,"time_ms":640,"url":"https://pubsub.googleapis.com/v1/projects/broad-dsde-dev/subscriptions/rawls-async-import-topic-sub-test:pull"}
rawls [INFO] [17:51:56.880] [rawls-akka.actor.default-dispatcher-35] o.b.d.r.m.AvroUpsertMonitorActor - received async upsert message: PubSubMessage(BCEhPjA-RVNEUAYWLF1GSFE3GQhoUQ5PXiM_NSAoRRAGCBQFfH1wQU91XV92aFENGXJ9YHJqXhdVARcGeQlRGgdoTm11JW4EEXp0ZHBoWxQBAk1Wd1pz8JLxn_GQU0sxIf7xyrdfet_MjulgZiU9XxJLLD5-PT1FQV5AEkw2CERJUytDCypYEU4,,Map(workspaceName -> avro-test2, workspaceNamespace -> abbbbbanu, userEmail -> ansinghfirecloud@gmail.com, upsertFile -> import-service-batchupsert-dev/cf26869a-145d-45b8-a53b-5b68f37a4df5.rawlsUpsert, jobId -> cf26869a-145d-45b8-a53b-5b68f37a4df5))
rawls [DEBUG] [17:51:56.975] [scala-execution-context-global-631] o.b.d.r.google.HttpGooglePubSubDAO - {"method":"POST","payload":{"ackIds":["ISE-MD5FU0RQBhYsXUZIUTcZCGhRDk9eIz81IChFEAEIFAV8fXBBXnVeXHJoUQ0Zcn1gcmpeF1UBEFUrW1EbB2hObXUlbgQRenRkdWNeFgIDQ1R2V3OYnIHvq_ePYBclSvHt3qhvM-qe2r5MZiU9XxJLLD5-PTNFQV5AEkw2BkRJUytDCypYEU4E"]},"statusCode":200,"time_ms":190,"url":"https://pubsub.googleapis.com/v1/projects/terra-arrow-dev/subscriptions/avro-import-topic-sub:acknowledge"}`

Reply:
Cool -- I learned something today :) Is this something you could do as a unit test?

Reply:
if i must

Reply:
actually, I'm not sure - I think this would require having a monitor that just does 1 monitor pass and then stops.